saleor/payment/gateways/braintree/__init__.py

```diff
@@ -136,128 +115,109 @@ def get_client_token(**connection_params: Dict) -> str:
     return client_token
 
 
-def authorize(
-        payment: Payment,
-        payment_token: str,
-        **connection_params: Dict) -> Tuple[Transaction, str]:
+def process_payment(
+        payment_information: Dict, **connection_params: Dict) -> Dict:
+    auth_resp = authorize(payment_information, **connection_params)
+    payment_information['token'] = auth_resp['transaction_id']
+
+    try:
+        assert auth_resp['is_success']
+        return [auth_resp, capture(payment_information, **connection_params)]
+    except Exception:
+        return [auth_resp, void(payment_information, **connection_params)]
+
+
+def authorize(payment_information: Dict, **connection_params: Dict) -> Dict:
     gateway = get_braintree_gateway(**connection_params)
+
     try:
         result = gateway.transaction.sale({
-            'amount': str(payment.total),
-            'payment_method_nonce': payment_token,
+            'amount': str(payment_information['amount']),
+            'payment_method_nonce': payment_information['token'],
             'options': {
                 'submit_for_settlement': CONFIRM_MANUALLY,
                 'three_d_secure': {
                     'required': THREE_D_SECURE_REQUIRED}},
-            **get_customer_data(payment)})
+            **get_customer_data(payment_information)})
     except braintree_sdk.exceptions.NotFoundError:
-        return transaction_and_incorrect_token_error(
-            payment, kind=TransactionKind.AUTH, token=payment_token)
+        raise BraintreeException
+
     gateway_response = extract_gateway_response(result)
     error = get_error_for_client(gateway_response['errors'])
-    credit_card_data = gateway_response.pop('credit_cart', None)
-    if credit_card_data:
-        payment.cc_first_digits = credit_card_data.bin
-        payment.cc_last_digits = credit_card_data.last_4
-        payment.cc_brand = credit_card_data.card_type
-        payment.cc_exp_month = credit_card_data.expiration_month
-        payment.cc_exp_year = credit_card_data.expiration_year
-        payment.save(update_fields=[
```

Comment:
What if we truly would like to update the payment from the gateway response? How can we do this with the new payment?

Reply:
With the new payment it's not possible for the gateway to update it. In the future if the need arises we can extend the functionality so the gateway can update the payment. For the time being only Saleor does any updating.